_posts/2021-04-05-maven-bom.md

```diff
@@ -0,0 +1,312 @@
+
+```markdown
++- org.springframework:spring-context:jar:5.3.5:compile
+|  +- org.springframework:spring-aop:jar:5.3.5:compile
+|  |  +- (org.springframework:spring-beans:jar:5.3.5:compile - omitted for duplicate)
+|  |  \- (org.springframework:spring-core:jar:5.3.5:compile - omitted for duplicate)
+|  +- org.springframework:spring-beans:jar:5.3.5:compile
+|  |  \- (org.springframework:spring-core:jar:5.3.5:compile - omitted for duplicate)
+...
+   +- (org.springframework:spring-expression:jar:5.2.13.RELEASE:compile - omitted for conflict with 5.3.5)
+   \- org.springframework:spring-web:jar:5.2.13.RELEASE:compile
+      +- (org.springframework:spring-beans:jar:5.2.13.RELEASE:compile - omitted for conflict with 5.3.5)
+      \- (org.springframework:spring-core:jar:5.2.13.RELEASE:compile - omitted for conflict with 5.3.5)
+
+```
+
+Maven indicates for each transitive dependency why it was omitted:
+
+- Omitted for duplicate if it's the same version in the same level.
+- Omitted for conflict if different versions in the same graph depth and not the first one in resolution order.
+
+Now it is clear for us how Maven resolves transitive dependencies. For some reason, we may be tempted one day to pick a specific version of a dependency and get rid of all the processes made by Maven to select it. To do this we have two options:
+
+- #### Use a direct dependency
+
+
+Adding the desired transitive dependency version as a direct dependency in the POM file will result in being the nearest in depth. This way Maven will select this version. In our previous example, if we wanted version `v1.3` to be selected, then adding the dependency `DT(v1.3)` in the POM file will ensure its selection.
+
+- #### Introduce `dependencyManagement` section in the POM file
+
+
+For projects with sub-modules, to ensure compatibility and coherence between all the modules, we need a way to provide the same version of a dependency across all sub-modules. Here comes in action the `dependencyManagement` section: it provides a lookup table for Maven to help determine the selected version of a transitive dependency and to centralize dependency information.
+
+A `dependencyManagement`  section contains dependency  elements. Each dependency is a lookup reference for Maven to determine the version to select for transitive dependencies. The version of the dependency is mandatory in this section. However, outside `dependencyManagement`  section, you can omit the version from your dependencies, and Maven will select the correct version of the transitive dependencies from the list of dependencies provided in `dependencyManagement`.
+
+We should note that defining a dependency in the `dependencyManagement` section doesn't add it to the dependency tree of the project, it is used just for lookup reference. Instead, for any dependency in a project's dependency tree, if it appears in the BOM file, then the version from the BOM will be used.
```

Comment:
The second sentence about the BOM file doesn't make sense, yet, because we haven't introduced the BOM file, yet. What do you mean with that? Or can we remove that sentence?

Reply:
I will remove it since we didn't introduce BOM yet.
I tries to say that Maven will resolve dependencies'versions from BOM if one is used.